### ISSUE TITLE
Command "brew update" shows "/dev/stdin: No such file or directory - /tmp/sh-thd-2853475760 (deleted) (Errno::ENOENT)"

### ISSUE THREAD

**Issue by azhongzl** [Homebrew/brew#17008]
- [ ] your problem was from running the official `install` or `uninstall` script? 
- [x] after installation: ran `brew config` and `brew doctor` and included their output with your issue? If you couldn't install: provided your OS version with the output of your issue?
```
brew config

HOMEBREW_VERSION: 4.2.16
ORIGIN: https://github.com/Homebrew/brew
HEAD: bd1155be8f50998429a795c15a69c8fe75250510
Last commit: 28 hours ago
Core tap JSON: 02 Apr 19:11 UTC
HOMEBREW_PREFIX: /home/linuxbrew/.linuxbrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: vim
HOMEBREW_MAKE_JOBS: 48
Homebrew Ruby: 3.1.4 => /home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: 48-core 64-bit zen3
Clang: N/A
Git: 2.40.1 => /bin/git
Curl: 8.3.0 => /bin/curl
Kernel: Linux 5.10.213-179.855.amzn2int.x86_64 x86_64 GNU/Linux
OS: Amazon Linux 2
Host glibc: 2.26
/usr/bin/gcc: 7.3.1
/usr/bin/ruby: 2.0.0
glibc: 2.35_1
gcc@11: N/A
gcc: 13.2.0
xorg: N/A
```

```
brew doctor

Your system is ready to brew.
```

## What you were trying to do (and why)

I am trying to run `brew update` to update brew.

## What happened (include command output)

Run below command
```
brew update
```

<details>
  <summary>Command output</summary>
  <pre>

brew update

/dev/stdin: No such file or directory - /tmp/sh-thd-2853451899 (deleted) (Errno::ENOENT)
==> Updating Homebrew...
Already up-to-date.

  </pre>
</details>

## What you expected to happen

Brew updates successfully without error like `/dev/stdin: No such file or directory - /tmp/sh-thd-2853451899 (deleted) (Errno::ENOENT)`

## Step-by-step reproduction instructions (by running `brew` commands)

```
brew update
```

**Comment by azhongzl** [comment]
1. ssh to the host
2. launch iTerm2
3. run command `brew update`

**Comment by MikeMcQuaid** [comment]
@azhongzl Can you reproduce on any other hosts or Linux machines? How long has this been happening? What user are you running as?

**Comment by azhongzl** [comment]
It worked before but just started to fail yesterday

**Comment by Bo98** [comment]
I'm guessing this is something weird with the way your system interacts with herestrings.

I'm guessing changing 

https://github.com/Homebrew/brew/blob/f0e442d88f5a4025ccd9d7c3bcf57a4cc3769baf/Library/Homebrew/utils/lock.sh#L50-L51

to be `echo "exit(RUBY_VERSION >= '1.8.7')" | "${ruby}" /dev/stdin` might fix things? And similarly on the line below.

**Comment by Bo98** [comment]
Maybe, though ideally we'd fix 2.0 otherwise we would need to bump the 1.8.7 requirement.

There's perhaps also an argument of hoisting the `flock` condition to the top: https://github.com/Homebrew/brew/blob/f0e442d88f5a4025ccd9d7c3bcf57a4cc3769baf/Library/Homebrew/utils/lock.sh#L56-L58
